include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h

```diff
@@ -1822,9 +1874,9 @@ namespace internal
 
         // if all access goes to the locally active dofs on all ranks, we do
         // not need the vec_fine vector
-        if (Utilities::MPI::max(static_cast<unsigned int>(
-                                  fine_indices_touch_non_active_dofs),
-                                comm) == 0)
+        if (false && Utilities::MPI::max(static_cast<unsigned int>(
+                                           fine_indices_touch_non_active_dofs),
+                                         comm) == 0)
```

Comment:
@kronbichler I had to disable this here, since the heuristic did not turn out to working well. See: `multigrid-global-coarsening/mg_transfer_p_01.mpirun=2.debug`.

Reply:
Can you be more specific in what sense it fails? The aim of the code is not a heuristic, but it should reflect the actual access pattern during the transfer operations 1:1. So we should find out what kinds of entries get forgotten for the present case, or if there is other uses of the fine vectors. This optimization is crucial and skipping typically costs 2x of performance in the transfer, as the whole thing is memory limited from the input/output vector, so doubling the bigger of the vectors should be limited to the case where one has to.

Reply:
The problem is that the newest changes implicitly lead to an internal partitioner that only contain locally active dofs also in the case of CG. All the indices are pre-computed for this partitioner. However, if an external vector is set up with locally relevant dofs the indices will point to the wrong entries. My guess is that we don't want to check here:

https://github.com/dealii/dealii/blob/d43a8d4667d0e05050f582868d232cb6a81baf6c/include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h#L1784-L1785

if the dofs are locally active but locally owned, since their position is guaranteed in the vectors in contrast to the ghost entries.

Reply:
Correct, we need to restrict the optimized case (for now) to the locally owned dofs. This is also why we need to pass in the partitioners to https://github.com/dealii/dealii/blob/d43a8d4667d0e05050f582868d232cb6a81baf6c/include/deal.II/multigrid/mg_transfer_matrix_free.h#L103-L106 as we can't know the numbers (now I remember why). We might (after the 9.3 release) want to add this mechanism also here.

Reply:
Moved #12213.